storage/integrate.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+
+	"github.com/transparency-dev/merkle/compact"
+	"github.com/transparency-dev/merkle/rfc6962"
+	"github.com/transparency-dev/trillian-tessera/api"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"golang.org/x/exp/maps"
+	"k8s.io/klog/v2"
+)
+
+// getPopulatedTileFunc is the signature of a function which can return a fully populated tile for the given tile coords.
+type getPopulatedTileFunc func(ctx context.Context, tileID TileID, treeSize uint64) (*populatedTile, error)
+
+// TreeBuilder constructs Merkle trees.
+//
+// This struct it indended to be used by storage implementations during the integration of entries into the log.
+// TreeBuilder caches data from tiles to speed things up, but has no mechanism for evicting from its internal cache,
+// so while it _may_ be possible to use the same instance across a number of integration runs (e.g. if the same job
+// is responsible for integrating entries for a number of contiguous trees), the lifetime should be bounded so as not
+// to leak memory.
+type TreeBuilder struct {
+	readCache *tileReadCache
+	getTile   getPopulatedTileFunc
+	rf        *compact.RangeFactory
+}
+
+// NewTreeBuilder creates a new instance of TreeBuilder.
+//
+// The getTile param must know how to fetch the specified tile from storage. It must return an instance of os.ErrNotExist
+// (either directly, or wrapped) if any of the the requested tiles were not found.
+func NewTreeBuilder(getTiles func(ctx context.Context, tileIDs []TileID, treeSize uint64) ([]*api.HashTile, error)) *TreeBuilder {
+	readCache := newTileReadCache(getTiles)
+	r := &TreeBuilder{
```

Comment:
This is missing an assignment to `getTile` in the `TreeBuilder`. Which probably suggests missing tests as CI hasn't caught any problems, and this field is used in `Integrate`.

Reply:
Whoops, that shouldn't be there - fixed.
Added a simple test for Integrate too.